firestore/lib/src/data/restaurant_provider.dart

```diff
@@ -0,0 +1,168 @@
+/*
+ *  Copyright 2022 Google LLC
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+import 'dart:async';
+
+import 'package:cloud_firestore/cloud_firestore.dart';
+
+import '../model/filter.dart';
+import '../model/restaurant.dart';
+import '../model/review.dart';
+
+abstract class RestaurantProvider {
+  Stream<List<Restaurant>> get allRestaurants;
+  Future<void> addRestaurant(Restaurant restaurant);
+  Future<void> addReview(
+      {required String restaurantId, required Review review});
+  void addRestaurantsBatch(List<Restaurant> restaurants);
+  void loadAllRestaurants();
+  void loadFilteredRestaurants(Filter filter);
+  Future<Restaurant> getRestaurantById(String restaurantId);
+  void dispose();
+}
+
+class FirestoreRestaurantProvider implements RestaurantProvider {
+  FirestoreRestaurantProvider() {
+    allRestaurants = _allRestaurantsController.stream;
+  }
+
+  final StreamController<List<Restaurant>> _allRestaurantsController =
+      StreamController();
+
+  @override
+  late final Stream<List<Restaurant>> allRestaurants;
+
+  @override
+  Future<void> addRestaurant(Restaurant restaurant) {
+    final restaurants = FirebaseFirestore.instance.collection('restaurants');
+    return restaurants.add({
+      'avgRating': restaurant.avgRating,
+      'category': restaurant.category,
+      'city': restaurant.city,
+      'name': restaurant.name,
+      'numRatings': restaurant.numRatings,
+      'photo': restaurant.photo,
+      'price': restaurant.price,
+    });
+  }
+
+  @override
+  void addRestaurantsBatch(List<Restaurant> restaurants) {
+    for (var restaurant in restaurants) addRestaurant(restaurant);
+  }
+
+  @override
+  Future<void> addReview(
+      {required String restaurantId, required Review review}) async {
+    final restaurant = await FirebaseFirestore.instance
+        .collection('restaurants')
+        .doc(restaurantId);
+    final newReview = await restaurant.collection('ratings').doc();
+
+    return FirebaseFirestore.instance
+        .runTransaction((Transaction transaction) async {
+      final DocumentSnapshot _restaurantSnapshot =
+          await transaction.get(restaurant);
+      final _restaurant = Restaurant.fromSnapshot(_restaurantSnapshot);
```

Comment:
nit: I have mixed feelings about this pattern (though it is very common!), because it forces our basic Restaurant models to know about Firestore documents, which further complicates other testing. A possible alternate is to introduce yet another layer, OR, put the method on the `FirestoreRestaurantProvider` class instead of as an extension on the data model itself.

I wouldn't overhaul everything right now, but I am interested in collaborating on how we much we want to publicly tackle this question in the future.

Reply:
fwiw, if you're anything like me and hadn't/haven't encountered much about domain driven design before, you may enjoy watching some of ResoCoder's content about it on YouTube. I know I did 😆